File: shared/origami/src/origami/comm/collective.cpp
/*******************************************************************************
 *
 * MIT License
 *
 * Copyright 2026 AMD ROCm(TM) Software
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 *******************************************************************************/

#include "origami/comm/collective.hpp"

namespace origami::comm {

double ring_step_overhead_cycles(primitive_t primitive,
                                 const collective_algorithm_t& algorithm,
                                 const heuristics_t& heur,
                                 const comm_hardware_t& fabric) {
  if (!algorithm.is_ring_class()) return 0.0;
  // The heuristic is host wall time (ns); convert to cycles at the actual GPU
  // clock. clock_ghz is cycles-per-ns, so cycles = ns × clock_ghz exactly.
  const double per_step_ns = heur.ring_step_overhead_ns_for(primitive);
  const double total_ns    = per_step_ns * static_cast<double>(algorithm.num_timesteps());
  return total_ns * fabric.clock_ghz;
}

double compute_ring_latency(const collective_algorithm_t& algorithm,
                            const comm_problem_t& problem,
                            const comm_config_t& config,
                            const system_t& system,
                            int my_rank,
                            const heuristics_t& heur) {
  const hardware_t& hw           = system.gpu;
  const comm_hardware_t& comm_hw = system.fabric;
  const int num_timesteps        = algorithm.num_timesteps();
  const std::size_t CL           = hw.cacheline_bytes;

  // A ring moves one chunk per step; over num_timesteps steps each GPU pushes
  // num_timesteps such chunks across its outgoing link. That product is the
  // total bytes this rank puts on the wire — the numerator of the throughput
  // model.
  const std::size_t gpu_timestep_tile_bytes =
      problem.gpu_tile_cachelines(hw.cacheline_bytes) * CL /
      static_cast<std::size_t>(algorithm.chunks_per_timestep());
  const std::size_t total_wire_bytes =
      gpu_timestep_tile_bytes * static_cast<std::size_t>(num_timesteps);

  const int eff_wgs = config.effective_num_wgs(gpu_timestep_tile_bytes);

  // Per-WG remote throughput is the same latency cap as in latency.hpp:
  // outstanding misses (mshr_depth × waves × CL) drained every RTT.
  const double mshr_bw_per_wg =
      (static_cast<double>(hw.mshr_depth_per_wave) * hw.waves_per_wg * static_cast<double>(CL)) /
      hw.xgmi_latency_cycles;

  // The ring's sustainable rate is whichever ceiling binds first: the physical
  // link width, or the combined latency-limited throughput of the WGs feeding
  // it (eff_wgs × per-WG cap). Few WGs -> CU-limited; many WGs -> link-limited.
  const double aggregate_bw =
      std::min(comm_hw.link_bw, static_cast<double>(eff_wgs) * mshr_bw_per_wg);

  const double T_transfer = static_cast<double>(total_wire_bytes) / aggregate_bw;

  // Sync: count signal_t+wait_t ops in the work graph times atomic_latency_cycles.
  const auto entry = algorithm.link_of(/*pid=*/0, /*timestep=*/0, my_rank);
  int sync_ops     = 0;
  for (const auto& op : entry.work_graph) {
    std::visit(
        [&](const auto& concrete) {
          using T = std::decay_t<decltype(concrete)>;
          if constexpr (std::is_same_v<T, signal_t> || std::is_same_v<T, wait_t>) { ++sync_ops; }
        },
        op);
  }
  // Each ring step needs its own handshake, and the steps are serialized by
  // the dependency chain, so sync cost accrues per step.
  const double T_sync_per_step = static_cast<double>(sync_ops) * comm_hw.atomic_latency_cycles;
  const double T_sync_total    = static_cast<double>(num_timesteps) * T_sync_per_step;

  // The same bytes that cross the fabric must also be read from / written to
  // local HBM, which has its own (CU-count-scaled) aggregate ceiling. The ring
  // can be bound by either resource, so take the slower of fabric and HBM.
  const double hbm_bw_agg = hw.hbm_read_bw * hw.bw_fraction(eff_wgs);
  const double T_hbm      = static_cast<double>(total_wire_bytes) / hbm_bw_agg;

  const double T_transfer_total = std::max(T_transfer, T_hbm);

  // Per-step proxy/handshake overhead the bandwidth model cannot see (CPU-
  // mediated; empirical, from heuristics). Keyed by the collective being run.
  const double T_step_overhead =
      ring_step_overhead_cycles(problem.collective, algorithm, heur, comm_hw);

  // Fixed launch floor + the throughput-bound transfer + serial sync + per-step
  // overhead. Launch dominates tiny messages; transfer dominates large ones.
  return comm_hw.launch_overhead_cycles + T_transfer_total + T_sync_total + T_step_overhead;
}

double compute_sequential_latency(const collective_algorithm_t& algorithm,
                                  const comm_problem_t& problem,
                                  const comm_config_t& config,
                                  const system_t& system,
                                  int my_rank,
                                  const heuristics_t& heur) {
  const hardware_t& hw           = system.gpu;
  const comm_hardware_t& comm_hw = system.fabric;
  const int chunks_per_timestep  = algorithm.chunks_per_timestep();
  // The collective being implemented; threaded down to the per-tile heuristics
  // (k_xgmi_write / ring_step_overhead) which take only the primitive, not the
  // full problem.
  const primitive_t primitive = problem.collective;

  const tile_shape_t gpu_tile = problem.gpu_tile_shape();
  const tile_shape_t gpu_timestep_tile =
      gpu_tile.divide_byte_equal(static_cast<std::size_t>(chunks_per_timestep));

  const int eff_wgs = config.effective_num_wgs(gpu_timestep_tile.bytes());

  const tile_shape_t wg_tile =
      gpu_timestep_tile.divide_byte_equal(static_cast<std::size_t>(eff_wgs));
  const wg_tile_geometry_t wg_geometry =
      wg_tile_geometry_t::from_shape(wg_tile, hw.cacheline_bytes);
  const latency_context_t lat_ctx{config, system, heur, primitive};

  // This loop is the iterative driver of the schedule: the algorithm methods are
  // closed-form (one timestep per call), and stepping through the timeline happens
  // here, not inside them. Timesteps are data-dependent, so their latencies add up.
  // Within a timestep, however, the links run in parallel — so a timestep costs the
  // *slowest* link, not their sum (T_link_max below).
  double T_timesteps = 0.0;
  for (int timestep = 0; timestep < algorithm.num_timesteps(); ++timestep) {
    const auto entry = algorithm.link_of(/*pid=*/0, timestep, my_rank);

    if (entry.is_self) {
      // A "self" step is a local copy (no peer): bound by local HBM, so the
      // per-WG budget is the HBM per-CU share rather than a link share.
      const double bw_per_wg = hw.hbm_read_bw_per_cu(eff_wgs);
      const auto breakdown   = compute_wg_tile_latency(entry.work_graph,
                                                       wg_geometry,
                                                       bw_per_wg,
                                                       /*active_cus=*/eff_wgs,
                                                       lat_ctx);
      T_timesteps += breakdown.T_total_cycles;
    } else {
      // A remote step may light up several links at once; wgs_per_active_link
      // reports how the eff_wgs workgroups are distributed over them (one count
      // per active link). Each link's WGs share that link's width evenly, and the
      // timestep waits for the most congested link to finish — hence the max over
      // links.
      const auto link_wg_counts = algorithm.wgs_per_active_link(timestep, eff_wgs);

      double T_link_max = 0.0;
      for (const int wgs_on_link : link_wg_counts) {
        const double bw_per_wg = comm_hw.link_bw / static_cast<double>(std::max(wgs_on_link, 1));

        const auto breakdown = compute_wg_tile_latency(entry.work_graph,
                                                       wg_geometry,
                                                       bw_per_wg,
                                                       /*active_cus=*/eff_wgs,
                                                       lat_ctx);
        T_link_max           = std::max(T_link_max, breakdown.T_total_cycles);
      }
      T_timesteps += T_link_max;
    }
  }

  const double T_step_overhead = ring_step_overhead_cycles(primitive, algorithm, heur, comm_hw);
  return comm_hw.launch_overhead_cycles + T_timesteps + T_step_overhead;
}

double compute_collective_latency_for_rank(const comm_problem_t& problem,
                                           const comm_config_t& config,
                                           const system_t& system,
                                           int my_rank,
                                           const heuristics_t& heur) {
  std::unique_ptr<collective_algorithm_t> owned;
  const collective_algorithm_t* A = config.algorithm_override;
  if (!A) {
    owned = resolve_algorithm(problem, config);
    A     = owned.get();
  }

  if (A->is_ring_pipeline()) {
    return compute_ring_latency(*A, problem, config, system, my_rank, heur);
  }
  return compute_sequential_latency(*A, problem, config, system, my_rank, heur);
}

double compute_collective_latency(const comm_problem_t& problem,
                                  const comm_config_t& config,
                                  const system_t& system,
                                  const heuristics_t& heur) {
  if (heur.assume_rank_symmetry) {
    return compute_collective_latency_for_rank(problem, config, system, /*my_rank=*/0, heur);
  }

  double T_max = 0.0;
  for (int rank = 0; rank < problem.num_gpus; ++rank) {
    T_max =
        std::max(T_max, compute_collective_latency_for_rank(problem, config, system, rank, heur));
  }
  return T_max;
}

double predict_row(std::string_view primitive,
                   std::size_t msg_bytes,
                   int world_size,
                   int nchannels,
                   const system_t& system,
                   std::size_t M,
                   std::size_t N,
                   int split_dim,
                   const heuristics_t& heur) {
  constexpr data_type_t dtype = data_type_t::BFloat16;

  if (M == 0 || N == 0) {
    const std::size_t total_elements = msg_bytes / static_cast<std::size_t>(dtype_bytes(dtype));
    std::size_t per_rank_elements;
    if (primitive == "reduce_scatter") {
      per_rank_elements =
          std::max<std::size_t>(total_elements / static_cast<std::size_t>(world_size), 1);
    } else {
      per_rank_elements = total_elements;
    }
    M = 1;
    N = per_rank_elements;
  }

  comm_problem_t problem{M, N, world_size, dtype, split_dim};
  problem.collective = primitive_from_name(primitive);  // string → enum at the edge
  comm_config_t config{};
  config.num_wgs          = nchannels;
  config.load_width       = load_width_t::DWORDX16;
  config.vgprs_for_data   = 128;
  config.min_bytes_per_wg = heur.min_bytes_per_wg;

  const double T_cycles = compute_collective_latency(problem, config, system, heur);

  return system.gpu.cycles_to_us(T_cycles);
}

}  // namespace origami::comm

File: model/collective.py
"""
End-to-end collective latency prediction.

Wires together: layout → contention → wg_tile cost → total latency.

Two computation modes based on layout structure:
1. Sequential timesteps: each timestep uses a (potentially) different link
   (e.g., one_shot AR visiting each peer in turn). T = Σ T_timestep.
2. Pipelined ring: all timesteps use the same link, but multiple WGs (channels)
   pipeline through ring steps. T = total_data / aggregate_throughput + sync.

Naming: gpu_tile → gpu_timestep_tile → wg_tile → iter_tile.
See model/types.py CommProblem docstring for the full hierarchy.
"""

from math import ceil
from typing import Optional

from .types import CommProblem, CommConfig, CACHELINE_BYTES, dtype_bytes, DataType
from .hardware import Hardware, CommHardware
from .heuristics import Heuristics, DEFAULT_HEURISTICS
from .layouts import (
    CollectiveLayout,
    RingAllReduceLayout,
    RingAllGatherLayout,
    RingReduceScatterLayout,
    RingFixedLayout,
    allgather_layout,
    reduce_scatter_layout,
    allreduce_one_shot_layout,
    allreduce_two_shot_layout,
    allreduce_ring_layout,
    alltoall_layout,
    broadcast_layout,
)
from .latency import compute_wg_tile_latency
from .primitives import Signal, Wait


# Ring-style layouts charged the per-step proxy/sync overhead heuristic.
# Non-ring layouts (TwoShot AR, A2A, PidStaggered) get 0 by default.
RING_LAYOUT_CLASSES = (
    RingAllReduceLayout,
    RingAllGatherLayout,
    RingReduceScatterLayout,
    RingFixedLayout,
)


def _ring_step_overhead_cycles(primitive: Optional[str], layout, heuristics: Heuristics) -> float:
    """Heuristic per-step proxy/sync overhead in cycles, summed over all steps.

    Returns 0 when the layout isn't ring-style or the primitive has no
    fitted overhead. See Heuristics.ring_step_overhead_cycles.
    """
    if not isinstance(layout, RING_LAYOUT_CLASSES):
        return 0.0
    if primitive is None:
        return 0.0
    per_step = heuristics.ring_step_overhead_cycles.get(primitive, 0.0)
    return per_step * layout.num_timesteps


COLLECTIVE_LAYOUTS = {
    "all_gather": allgather_layout,
    "reduce_scatter": reduce_scatter_layout,
    "all_reduce": allreduce_two_shot_layout,  # pid-staggered: channels spread across peers
    "all_to_all": alltoall_layout,
    "broadcast": broadcast_layout,
}


def _is_ring_layout(layout):
    return isinstance(layout, (RingAllReduceLayout, RingFixedLayout))


def compute_collective_latency(
    collective: str,
    problem: CommProblem,
    config: CommConfig,
    hw: Hardware,
    comm_hw: CommHardware,
    layout: Optional[CollectiveLayout] = None,
    my_rank: int = 0,
    heuristics: Heuristics = DEFAULT_HEURISTICS,
) -> float:
    """
    Predict total collective latency in **GPU cycles**.

    The conversion to seconds happens once at the top of the model
    (`predict_row`), dividing by `hw.clock_ghz * 1000` to produce µs.

    `heuristics` supplies xGMI write concentration and per-ring-step
    overhead constants (see Heuristics).
    """
    if layout is None:
        layout_fn = COLLECTIVE_LAYOUTS.get(collective)
        if layout_fn is None:
            raise ValueError(f"Unknown collective: {collective}")
        layout = layout_fn(problem.num_gpus)

    if _is_ring_layout(layout):
        return _compute_ring_latency(layout, problem, config, hw, comm_hw, my_rank, heuristics, collective)
    else:
        return _compute_sequential_latency(layout, problem, config, hw, comm_hw, my_rank, heuristics, collective)


def _compute_ring_latency(
    layout, problem, config, hw, comm_hw, my_rank,
    heuristics: Heuristics = DEFAULT_HEURISTICS,
    primitive: Optional[str] = None,
) -> float:
    """
    Ring latency: all timesteps use the same link. Multiple WGs (channels)
    pipeline through the ring steps, so the aggregate throughput approaches
    link bandwidth when enough WGs are active.

    Total wire data crossing the link = gpu_timestep_tile_bytes × num_timesteps.
    Each WG handles gpu_timestep_tile_bytes / effective_num_wgs bytes per step.
    The link is shared by `effective_num_wgs` WGs simultaneously — over-launched
    channels (more WGs than the timestep can feed) are idle for aggregate
    throughput purposes (see `CommConfig.effective_num_wgs`).

    Aggregate link throughput = min(link_bw, effective_num_wgs × mshr_bw_per_wg).
    T_transfer = total_wire_data / aggregate_throughput.
    T_sync = num_timesteps × per_step_sync (sync still costs full num_wgs).
    """
    N = problem.num_gpus
    num_wgs = config.num_wgs
    num_timesteps = layout.num_timesteps  # 2(N-1) for AllReduce, N-1 for AG/RS

    # Per-timestep data crossing the link per GPU (not per WG).
    # NOTE: with chunks_per_timestep=1 (the current default), this equals the
    # whole gpu_tile per timestep — behaviorally preserved from the pre-rename
    # code. The bug fix is to set chunks_per_timestep=N on chunked layouts and
    # let the division below shrink each timestep's data accordingly.
    gpu_timestep_tile_bytes = (
        problem.gpu_tile_cachelines * CACHELINE_BYTES // layout.chunks_per_timestep
    )
    total_wire_bytes = gpu_timestep_tile_bytes * num_timesteps

    # Cap WGs to the amount of work actually available per timestep — see
    # CommConfig.effective_num_wgs and `min_bytes_per_wg`.
    eff_wgs = config.effective_num_wgs(gpu_timestep_tile_bytes)

    # Per-WG MSHR-limited throughput (bytes/cycle)
    CL = CACHELINE_BYTES
    mshr_bw_per_wg = (hw.mshr_depth_per_wave * hw.waves_per_wg * CL) / hw.xgmi_latency_cycles

    # Aggregate throughput: only ACTIVE WGs contribute to throughput, but the
    # per-link physical cap is independent of WG count.
    aggregate_bw = min(
        comm_hw.link_bw,                    # per-link physical cap
        eff_wgs * mshr_bw_per_wg,           # CU-limited aggregate (active only)
    )

    T_transfer = total_wire_bytes / aggregate_bw

    # Per-timestep sync: each of the num_timesteps logical steps has signal/wait
    # overhead. With pipelining, sync is overlapped. Only the first and last
    # steps aren't overlapped. Model as: num_timesteps × sync_per_step, where
    # sync_per_step is the atomic RTT (not the full step time, since sync
    # overlaps with transfer).
    entry = layout.link_of(0, 0, my_rank, N)
    sync_ops = sum(1 for op in entry.work_graph if isinstance(op, (Signal, Wait)))
    T_sync_per_step = sync_ops * comm_hw.atomic_latency_cycles
    T_sync_total = num_timesteps * T_sync_per_step

    # Also account for local HBM read/write (Load + Store in work graph).
    # For RS phase: each step reads msg/N from local HBM and writes msg/N.
    # For AG phase: each step writes msg/N.
    # This traffic shares HBM BW with the xGMI traffic.
    # At low WG count, HBM per-CU BW may be the bottleneck.
    # Use `eff_wgs` so HBM polynomial scaling reflects real consumer count.
    hbm_bw_agg = hw.hbm_read_bw * hw._bw_fraction(eff_wgs)
    T_hbm = total_wire_bytes / hbm_bw_agg  # local read side

    T_transfer_total = max(T_transfer, T_hbm)

    # Per-ring-step proxy/sync overhead (heuristic). Additive to T_sync_total
    # (which covers in-graph atomic RTTs from Signal/Wait ops).
    T_step_overhead = _ring_step_overhead_cycles(primitive, layout, heuristics)

    return comm_hw.launch_overhead_cycles + T_transfer_total + T_sync_total + T_step_overhead


def _compute_sequential_latency(
    layout, problem, config, hw, comm_hw, my_rank,
    heuristics: Heuristics = DEFAULT_HEURISTICS,
    primitive: Optional[str] = None,
) -> float:
    """
    Sequential timesteps: each timestep uses a (potentially) different link.
    T = Σ T_timestep across all timesteps.
    """
    N = problem.num_gpus
    num_wgs = config.num_wgs

    # 2D tile hierarchy. Each level is a TileShape (m × n + dtype + split_dim)
    # so that downstream consumers (cacheline-padding accounting, future
    # heterogeneous gemm↔comm interop) can keep the shape; the scalar
    # `cachelines` / `elements` values fed into `compute_wg_tile_latency` are
    # just projections of the 2D tile.
    chunks_per_timestep = layout.chunks_per_timestep
    gpu_tile = problem.gpu_tile_shape
    gpu_timestep_tile = gpu_tile.divide_byte_equal(chunks_per_timestep)

    # Cap WG count by available data — over-launched WGs sit idle and don't
    # contend for HBM/L2/xGMI bandwidth. See `CommConfig.effective_num_wgs`.
    eff_wgs = config.effective_num_wgs(gpu_timestep_tile.bytes)

    wg_tile = gpu_timestep_tile.divide_byte_equal(eff_wgs)
    wg_tile_cachelines = max(wg_tile.cachelines, 1)
    wg_tile_elements = max(wg_tile.elements, 1)

    T_timesteps = 0.0
    for timestep in range(layout.num_timesteps):
        entry = layout.link_of(0, timestep, my_rank, N)

        if entry.is_self:
            # Self-timestep: local HBM only, no xGMI. Per-CU HBM BW uses the
            # effective WG count so that low-active-CU regimes get the full
            # per-CU share rather than a polynomial-discounted slice.
            bw_per_wg = hw.hbm_read_bw_per_cu(eff_wgs)
            breakdown = compute_wg_tile_latency(
                work_graph=entry.work_graph,
                wg_tile_cachelines=wg_tile_cachelines,
                config=config,
                hw=hw,
                comm_hw=comm_hw,
                bw_per_wg=bw_per_wg,
                wg_tile_elements=wg_tile_elements,
                wg_tile=wg_tile,
                active_cus=eff_wgs,
                heuristics=heuristics,
                primitive=primitive,
            )
            T_timesteps += breakdown.T_total_cycles
        else:
            # Layout distributes ACTIVE WGs across links — over-launched
            # idle WGs don't share link BW.
            link_wg_counts = layout.active_links(timestep, eff_wgs, N)

            T_links = []
            for link_id, wgs_on_link in link_wg_counts.items():
                bw_per_wg = comm_hw.link_bw / max(wgs_on_link, 1)

                breakdown = compute_wg_tile_latency(
                    work_graph=entry.work_graph,
                    wg_tile_cachelines=wg_tile_cachelines,
                    config=config,
                    hw=hw,
                    comm_hw=comm_hw,
                    bw_per_wg=bw_per_wg,
                    wg_tile_elements=wg_tile_elements,
                    wg_tile=wg_tile,
                    active_cus=eff_wgs,
                    heuristics=heuristics,
                    primitive=primitive,
                )
                T_links.append(breakdown.T_total_cycles)

            T_timesteps += max(T_links) if T_links else 0.0

    # Per-ring-step proxy/sync overhead (heuristic). Zero for non-ring layouts.
    T_step_overhead = _ring_step_overhead_cycles(primitive, layout, heuristics)

    return comm_hw.launch_overhead_cycles + T_timesteps + T_step_overhead


def predict_row(
    primitive: str,
    msg_bytes: int,
    world_size: int,
    nchannels: int,
    hw: Hardware,
    comm_hw: CommHardware,
    M: int = 0,
    N: int = 0,
    split_dim: int = 0,
    heuristics: Heuristics = DEFAULT_HEURISTICS,
) -> float:
    """Predict latency in microseconds for one row of rccl_master_sweep.csv.

    If M and N are not provided, assumes 1D contiguous layout
    (M=1, N=per_rank_elements). `msg_bytes` follows the RCCL/NCCL CSV
    convention, which is per-op:
        AR / AG / BC / A2A  →  `msg_bytes` is the per-rank tile size
        RS                  →  `msg_bytes` is the *total input* (= W × per-rank)

    The internal `gpu_tile` is normalized to the per-rank tile for every op,
    so RS's `msg_bytes` is divided by `world_size` here. This keeps the
    contract uniform: `gpu_tile` always == per-rank, `chunks_per_timestep`
    handles the algorithm's internal subdivision.
    """
    dtype = DataType.BF16
    if M == 0 or N == 0:
        total_elements = msg_bytes // dtype_bytes(dtype)
        if primitive == "reduce_scatter":
            # CSV/NCCL convention: msg_bytes for RS is the total input across
            # all ranks; the per-rank tile is total / W.
            per_rank_elements = max(total_elements // world_size, 1)
        else:
            per_rank_elements = total_elements
        M = 1
        N = per_rank_elements

    problem = CommProblem(M=M, N=N, num_gpus=world_size, dtype=dtype, split_dim=split_dim)
    config = CommConfig(num_wgs=nchannels)

    # `compute_collective_latency` returns GPU cycles. The cycles → time
    # boundary is governed by the GPU frequency: seconds = cycles / clock_hz.
    # `hw.cycles_to_us` applies this conversion and scales to microseconds.
    T_cycles = compute_collective_latency(
        primitive, problem, config, hw, comm_hw, heuristics=heuristics,
    )
    return hw.cycles_to_us(T_cycles)
